### DISCUSSION TITLE
Disable the `prefer-named-capture-group` rule

### DISCUSSION THREAD

**Discussion by leroykorterink** [mediamonks/eslint-config#226] [General]
There is a proposal to disable the `prefer-named-capture-group` rule.

> `prefer-named-capture-group`: regex is already hard to read, this rule makes it even harder.
> \- @edgardz

**Comment by leroykorterink** [comment]
With the landing of ECMAScript 2018, named capture groups can be used in regular expressions, which can improve their readability. This rule is aimed at using named capture groups instead of numbered capture groups in regular expressions:

```ts
const regex = /(?<year>[0-9]{4})/;
```

Alternatively, if your intention is not to capture the results, but only express the alternative, use a non-capturing group:

```ts
const regex = /(?:cauli|sun)flower/;
```

Please see [the documentation](https://eslint.org/docs/latest/rules/prefer-named-capture-group) of the rule.

**Comment by ThaNarie** [comment]
> regex is already hard to read, this rule makes it even harder

Funny, I have the exact opposite :)

Named capture groups provide a way to document your regular expression inline.

Similar to the case where clear variable names don't require comments, using capture groups make regular expressions mostly self-documenting.

```ts
// I can clearly understand that this regexp parses year, month and day out of a date string, using that specific order.
const date = /(?<year>[0-9]{4})-(?<month>[0-9]{2})-(?<day>[0-9]{2})/u;
```

At the same time, using the regex result in your code becomes way clearer.

```ts
foo.exec('bar').groups.id;
// vs
foo.exec('bar')[1];
```

And you don't have to change your index if you add/remove a capturing groups in the middle.

Not using capture groups feels like writing functions where you always use `...args` as the parameter, and access the params via `args[0]`, `args[1]`, etc.

---

Additionally, forcing groups that you are not using the result of, to be **non-capturing groups**, result in better (and faster?) regular expressions, as people often abuse capturing groups just for modifiers, without event needing the result, resulting in unused "index gaps" in the results.
